(TK-148) Expose selectors and acceptors and adjust thread defaults

camlow325 opened the pull request:
This commit contains three primary changes:

1) Expose the ability to configure the number of `selector-threads` and
   `acceptor-threads` that the webserver will spin up per "port" in a
   server configuration.

2) Remove the library-imposed defaults for various thread-pool related
   settings, including: acceptor-threads, selector-threads, max-threads,
   and queue-max-size.  The service will now defer to Jetty to impose
   its own default values for these settings.

3) In the process of removing library-imposed defaults, this commit
   effectively removes the changes from TK-130.  The `max-threads` setting
   will no longer automatically be bumped up above the minimum number of
   threads that the server needs to boot if the number that the server
   tries to use is insufficient and the user had not configured an explicit
   value.  Some documentation was added to the `max-threads` setting to
   advise users on how to configure the value manually, if needed to avoid
   this error.

camlow325 commented on src/puppetlabs/trapperkeeper/services/webserver/jetty9_config.clj:
```diff
@@ -81,22 +81,6 @@
 ;;; leaving them on by default.
 (def default-jmx-enable "true")
 
-
```
These were all removed since we're deferring to Jetty's defaults for these now.

camlow325 commented on src/puppetlabs/trapperkeeper/services/webserver/jetty9_config.clj:
```diff
@@ -433,59 +427,6 @@
     (throw (IllegalArgumentException.
              "Error: More than one default server specified in configuration"))))
 
-(defn selectors-count
```
Production code wouldn't use these anymore however the count functions are still being used in the tests in the jetty9-default-config-test namespace to help validate that Jetty is using the algorithm that we think it is for calculating acceptor and selector threads.

camlow325 commented on src/puppetlabs/trapperkeeper/services/webserver/jetty9_core.clj:
```diff
@@ -270,25 +273,34 @@
+                        (QueuedThreadPool. max-threads)
+                        (QueuedThreadPool.))]
+      (if queue-max-size
+        (let [min-threads    (.getMinThreads thread-pool)
```
I hated having to do the double-construction of a `QueuedThreadPool` but couldn't think of a better way to configure Jetty to use a custom value for `queue-max-size` while still preserving the other defaults that Jetty uses for other settings in the thread pool configuration.

cprice404 replied:
Hmm.. so this is just for 'getMaxThreads' and 'getIdleTimeout'?  Do either of those involve any fancy computations or are they just hard-coded numeric values in Jetty?

If the latter, I'm thinking maybe it would be a lesser evil to just put those hard coded values up into our defaults section with some comments explaining why they're there, and a reference to the tests that validate them... and then a comment next to those tests explaining that if they fail in the future then the user needs to update the defaults?  Or maybe the tests should just be written to use those hard-coded vars in their implementation?

I know that goes against the grain of what we've been trying to do here, but it seems maybe slightly less bad than having to construct a thread pool just to throw it away?

camlow325 replied:
Most of it is just hard-coded numbers in Jetty - see https://github.com/eclipse/jetty.project/blob/jetty-9.2.10.v20150310/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java#L65-L98.  The only part that's somewhat more interesting is the max (_minThreads, 8) part, which the line below this in the Clojure code mimics.  The assumption being that the "8" hard-coded in two places in the Jetty code is representative of an intention to tie min-threads to the initial capacity of the thread pool's queue.

I'm not too worried about the expense of the double-construction.  Appears to mostly be setting up primitive values for fields in the `QueuedThreadPool` object.  The backing `BlockingArrayQueue` object will allocate an Object array per the initial capacity of the queue, which won't be larger than 8 - see https://github.com/eclipse/jetty.project/blob/jetty-9.2.10.v20150310/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java#L113-L118.  The expensive operations wouldn't happen until after the thread pool is started, which it never would be for the first one allocated here.

I think I'd still prefer the double-construction to hard coding defaults in our code since this is a one-time thing at startup and is pretty inexpensive.

cprice404 replied:
If you're sure it doesn't actually start creating threads then I don't have too big of a problem with this implementation, depending on what other folks think.

If we do keep it this way I think you should add a comment though, so that people don't think you're nuts when they look at this in the future.

camlow325 replied:
Yeah, none of the threads would start unless `start` were called on the `QueuedThreadPool` - https://github.com/eclipse/jetty.project/blob/jetty-9.2.10.v20150310/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java#L106 - which would never happen on the first one constructed, in this case.

I'll add a comment to the code, but I guarantee people will think I'm nuts no matter what.  :smile:

camlow325 commented on src/puppetlabs/trapperkeeper/services/webserver/jetty9_config.clj:
```diff
@@ -495,9 +436,11 @@
   (let [result (-> {}
                    (maybe-add-http-connector config)
                    (maybe-add-https-connector config)
-                   (assoc :max-threads (determine-max-threads config (num-cpus)))
```
We would no longer be auto-adjusting `max-threads` up to the needed server minimum - prior work from TK-130.  The assumption is that the user would now have to set `max-threads` manually in the case that the number of acceptor and selector threads that the server needs to use would exceed the default `max-threads`.  This probably wouldn't be common anyway but is now documented in the `max-threads` section of the Jetty config doc.

camlow325 commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_config_test.clj:
```diff
@@ -372,129 +409,6 @@
       "./dev-resources/config/jetty/ssl/certs/master-with-intermediate-ca.pem"
       "./dev-resources/config/jetty/ssl/certs/ca-root.pem")))
 
-(deftest determine-max-threads-test
```
These are all gone now because they were only needed to validate the auto-adjustment of `max-threads` from TK-130, which would no longer exist per this PR.

camlow325 commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_core_test.clj:
```diff
@@ -189,125 +192,139 @@
-   :max-threads max-threads
-   :queue-max-size queue-max-size})
-
-(defn get-thread-pool-and-queue-for-create-server
```
The guts of this moved over to `get-server-thread-pool-queue` in the jetty9-default-config-test namespace, since it now has a similar need for this.

camlow325 commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_core_test.clj:
```diff
@@ -189,125 +192,139 @@
-    {:thread-pool thread-pool
-     :queue queue}))
-
-(deftest create-server-thread-pool-test
```
These tests still effectively exist under `create-server-queue-max-size-test` but they've been refactored to eliminate some duplication and also validate Jetty defaults when not explicitly set in user configuration.

cprice404 commented:
Overall I'm a big +1 on this.  Lots of deletions.  yay :)

I only reviewed the tests at a high level, it's a bit tricky to review since stuff is moving around, but it generally looks good to me.

camlow325 commented:
I added a comment about the extra `QueuedThreadPool` construction for the `queue-max-size` case.  That's in commit 2249af4.

KevinCorcoran commented on src/puppetlabs/trapperkeeper/services/webserver/jetty9_core.clj:
```diff
@@ -270,25 +273,50 @@
+                             (BlockingArrayQueue.
+                               queue-min-size
+                               queue-min-size
+                               queue-max-size)))
```
I'm okay with the approach you took here (creating the extra thread pool) and your comment is helpful.  I think these changes are fine, however, I want to raise a question unrelated to your changes ... 

Why are we setting the min size and growth raise on the queue?  It seems like all we really care about is the max size, and there is a constructor that takes only the max size which we could use instead.  Furthermore, the calculation for min size doesn't quite make sense, and it's also quite confusing to me that we're setting growth rate == min size.

camlow325 replied:
If we go with the `BlockingArrayQueue` constructor that just takes a max capacity, we'd end up with a fixed-sized queue that's pre-allocated up to the maximum capacity specified.  See https://github.com/eclipse/jetty.project/blob/jetty-9.2.10.v20150310/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java#L101.

The Server's default `QueuedThreadPool`, however, creates a `BlockingArrayQueue` that is allocated with a minimum capacity with the ability to grow up toward the maximum.  See: https://github.com/eclipse/jetty.project/blob/jetty-9.2.10.v20150310/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java#L92-L96.  I was thinking we'd want to preserve the "grow" capability to keep memory usage in check and behave more like the way the server would by default.

`QueuedThreadPool` sets the min size and growth rate based on the maximum of min threads in the thread pool or "8".  The "8" here also corresponds to the default value for min threads in the pool.  Since we don't allow for min threads in the pool to be configured, I just went with fixing this to min threads in the pool only.

The `(min queue-max-size min-threads)` is something Jetty doesn't protect for but that seemed worthwhile.  `BlockingArrayQueue` will throw an IllegalArgumentException if the initial capacity is larger than the maximum capacity - see https://github.com/eclipse/jetty.project/blob/jetty-9.2.10.v20150310/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java#L133.  Since we don't expose a way to configure initial capacity, it seemed reasonable to me to just head off the possibility of the IllegalArgumentException by adjusting the initial capacity to be equal to maximum capacity.  I suspect this to be pretty unlikely anyway to set such a small max capacity for the queue for anything but testing.

KevinCorcoran replied:
Okay, I'm on-board with setting min size.  

However, in the second link you provided above, notice that the growth rate is not specified - so I'm still not sure why we are setting that, and why we're using `queue-min-size` as the value.

Also, this seems like really good information to have in (yet another) comment in this function.

camlow325 replied:
> However, in the second link you provided above, notice that the growth rate is not specified - so I'm still not sure why we are setting that, and why we're using queue-min-size as the value.

Not sure I follow.  Jetty's `QueuedThreadPool`, by default, is calling the form of the `BlockingArrayQueue` constructor that takes in a "capacity" and a "growBy".  See https://github.com/eclipse/jetty.project/blob/jetty-9.2.10.v20150310/jetty-util/src/main/java/org/eclipse/jetty/util/BlockingArrayQueue.java#L113.  It sets them both to the same value, which derives from the minimum number of threads in the `QueuedThreadPool`.

I could define a `queue-capacity` and `queue-grow-by` and set them both to the result of `(min queue-max-size min-threads)` if that would help for clarity, although that wouldn't produce a functional difference.

KevinCorcoran replied:
Oh, my bad, I was mistaken on what those 2 constructor params were.

I think that change would be an improvement, but I'm fine without it too - however, a comment would still be really helpful.

And thanks for talking me through this.

KevinCorcoran commented on doc/jetty-config.md:
```diff
@@ -16,21 +16,60 @@ This sets what port to use for _unencrypted_ HTTP traffic.  If not supplied, but
+Note that each web request must be processed on a "worker" thread which is
+separate from the acceptor and selector threads.  The configured `max-threads`
+value should allow for the maximum number of requests which are desired for the
+server be able to handle concurrently.
```
I think this last sentence needs re-wording, it's hard to understand as written.

Also, crazy question here, but, since we're making acceptor + selector thread counts configurable, should we do the same for the worker threads?

camlow325 replied:
Worker threads effectively are configurable.  Basically, the difference between the sums of `acceptor-threads` and `selector-threads` across connectors and `max-threads` is the maximum number of worker threads that the server will allocate.  Worker threads are different from `acceptor-threads` and `selector-threads` in they aren't fixed at boot time.  The server will only allocate one by default but can allocate additional threads, as needed to handle incoming traffic, up to the `max-threads` for the pool.  The server can also let the number of worker threads scale back down after a peak period for requests has passed.

If we hadn't already had the concept of a `max-threads` setting, it might have made more sense to me to have a `worker-threads` setting and we could use that to calculate the `max-threads` that we give to Jetty during configuration.  It might be too confusing to change that now, and it also would seem to get us into the mode of having our production code be more knowledgeable about exactly how this algorithm is used by Jetty - which I think we're trying to avoid as much as possible now.

KevinCorcoran commented on doc/jetty-config.md:
```diff
@@ -16,21 +16,60 @@ This sets what port to use for _unencrypted_ HTTP traffic.  If not supplied, but
+~~~~
+
+"1" is the minimum number of worker threads required to process incoming
+web requests.
```
I think you should move this sentence into the last paragraph where you are also talking about worker threads.

KevinCorcoran commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj:
```diff
@@ -71,19 +75,110 @@ react accordingly."
+                               " \\+ selectors="
+                               (* selector-thread-count connectors)
+                               " \\+ request=1\\)"))))]
+    (dotimes [x 5]
```
why 5 times?

camlow325 replied:
It's arbitrary.  Could just do 2 times to cover the maximum number we'd ever use for a connector, for now.  Our testing here is indirect anyway since we're throwing some numbers and Jetty and using our expectation of what it will calculate to give us some confidence that we know how their algorithm works.  I'd thought that by throwing a few more connectors in there that it might increase that confidence a bit more.  I could be convinced to bump it back down to 2, though, if you think that's better.

KevinCorcoran replied:
I just realized that the purpose of this test is that it will fail if Jetty changes the way the calculate the minimum required # of threads ... right?

In that case ... I'm not sure why repeating this more than once makes sense.  What could possibly cause it to pass a couple of times and then eventually fail?  Like, if Jetty changed something that made the calculation different on subsequent connectors?

I gotta say, 2 seems a lot less arbitrary than 5, so I'd feel better about that.   Also I think this test could use some extra comments or `testing` strings or whatever that capture some of this info.

camlow325 replied:
I'll go with 2 and add a comment that we're doing that because 2 is the maximum number of connectors our library allows to be associated with a server - at most one for plaintext and one for SSL.

KevinCorcoran commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj:
```diff
@@ -71,19 +75,110 @@ react accordingly."
+                             (.getClass)
+                             (.getDeclaredMethod "getQueue" nil))
+        _                (.setAccessible get-queue-method true)]
+    (.invoke get-queue-method thread-pool nil)))
```
this is pretty evil!  :clap:

KevinCorcoran commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_core_test.clj:
```diff
@@ -189,125 +192,139 @@
+                                {:queue-max-size 393})))))
+    (testing (str "default max threads passed through to thread pool when "
+                  "queue-max-size set")
+      (is (= (default-server-max-threads)
```
This looks like a bug - looks like accidental parens around `default-server-min-threads` - that seems like it should throw an error since `default-server-min-threads` should be an integer, right?

KevinCorcoran replied:
oh, nevermind

KevinCorcoran commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_core_test.clj:
```diff
@@ -189,125 +192,139 @@
+
+(defn default-server-max-threads
+  []
+  (.getMaxThreads (get-thread-pool-for-default-server)))
```
This stuff would probably be a bit more readable like...
```clj
(def default-server (Server.)))

(def default-server-max-threads (.getMaxThreads default-server)))
```

camlow325 replied:
Ok, I can do that.

KevinCorcoran commented on test/clj/puppetlabs/trapperkeeper/services/webserver/jetty9_default_config_test.clj:
```diff
@@ -71,19 +75,110 @@ react accordingly."
+(deftest default-min-threads-settings-test
+  ;; This test just exists to validate the advice we give for the bare
+  ;; minimum number of threads that one should account for when setting the
+  ;; 'max-threads' setting for a server instance.
```
Which advice is this referring to?  I just re-read the docs for `max-threads` but I couldn't figure out what you meant by that.

KevinCorcoran replied:
Oh, I guess you're just referring to the calculation in Jetty.

camlow325 replied:
Yeah, just the fact that Jetty calculates its min-threads based on the number of selector and acceptor threads that it needs to allocate.

camlow325 commented:
I added commit 9211f47 to address some PR feedback:

1) Hopefully clarified the language in the `max-threads` setting definition.

2) Added some more comments around the construction of the `QueuedThreadPool` when `queue-max-size` is set.

3) Changed a couple of defns to defs for stuff that could be memoized.

4) Changed the `default-min-threads-settings-test` to only exercise one and two connector per server scenarios since those are the only ones which are possible for a tk-jetty9 server.

KevinCorcoran commented:
Tests passed for me locally.
